### ISSUE TITLE
n_components>2 gives poor results

### ISSUE THREAD

**Issue by alexwilson1** [hyhuang00/ParamRepulsor#7]
Reducing data to n_components=3 gives what is almost a 2D plane:

![Image](https://github.com/user-attachments/assets/91a5b195-d9a3-49ba-a154-aff29f0fe3ed)

![Image](https://github.com/user-attachments/assets/8d75f732-62dc-47e7-9e84-32f816b9dcd0)

Whereas UMAP with the same n_components and n_neighbors gives something like:

![Image](https://github.com/user-attachments/assets/61be68af-0aa2-4317-b8e8-cb24116be7ab)

**Comment by williamsyy** [comment]
@alexwilson1 Thank you for your interest in ParamRepulsor. Can you provide us the code or data for us to reproduce your results? We haven't observed this circumstance before. Thank you!

**Comment by alexwilson1** [comment]
@williamsyy wondering if there are any updates on this one?

**Comment by williamsyy** [comment]
@alexwilson1 Sorry for the late reply. Based on our current findings, we think that the learning rate for ParamRepulsor might be one of the key reasons why the 2D plane has occurred. Based on the dataset that you have provided, if you set up the learning rate to 5e-2 instead of the default value, then it would avoid leading to the 2D plane structure in the 3D space.

![Image](https://github.com/user-attachments/assets/187def0e-635a-4341-b480-b126b4176489)

**Comment by alexwilson1** [comment]
interesting, thank you for the investigation and great findings! Do you have a heuristic for how we could choose this learning rate for our datasets more generally please?

**Comment by williamsyy** [comment]
Since the dataset size and complexity are unknown, I have to say it might be hard to tell how we should fine-tune the learning rate to get a good performance. The issue is mostly from the NN training strategies since the backbone of the ParamRepulsor is an ANN. Therefore, any kind of strategy for setting up the learning rate should work here. Based on our current trail, from the dataset size perspective, if your dataset is small, it would be better for you to increase the learning rate a bit so that it can better learn the manifold, and if your dataset is large, then you should probably decrease the learning rate. From a heuristic tunning perspective, if you start from the default learning rate, and just fine tune it slightly around the default value, then you are more likely to find a proper learning rate that works for 3D projection.

**Comment by alexwilson1** [comment]
@williamsyy Thank you for the detailed explanation! Your insight on how the dataset size and complexity affect the learning rate is really helpful. I'll experiment by increasing the learning rate for smaller datasets and decreasing it for larger ones, starting from the default value and fine-tuning from there. I appreciate your guidance on this—I'll keep you posted on how these adjustments work out in practice. Thanks again for your support!
